### ISSUE TITLE
Issue with finetuned Arcface model

### ISSUE THREAD

**Issue by zul96a** [serengil/deepface#350]
Hi again!

Referring to the title, I manage to retrain Arcface model using [Keras Insightface](https://github.com/leondgarse/Keras_insightface) implementation. I looked at arcface model from this repo and based on the model.summary(), it uses Resnet34 as base model. 

Following https://github.com/leondgarse/Keras_insightface/discussions/36 method on training Resnet34 model, I use CASIA webface plus another 4200 unique person from my database and I got accuracy of around 0.99% on lfw , 0.89% for agedb_30.

When I tried to replace the newly trained model, h5 , with the current arface_weights.h5, I see the results vary too much. For example before this, a true negative reading (cosine distance) between Image A and B was 0.89 but after changing the weight, it become 0.11 which is false positive. In your repo model,  set the threshold to be around 0.51 which anything below it is match

I suspect that the newly trained model structure might differ a bit from this repo version since the size in Bytes of the new model is 133,947 KB , instead of this repo, 133, 816. I also notice that from the Keras Insightface repo, they use a custom NormDense layer, so maybe this might be the issue? Or it could be tensorflow version?

I'm not sure what I can do at the moment, but I appreciate any insights that you can give me. Thank you!

**Comment by serengil** [comment]
you should tune the threshold value in that case. I mean that 0.11 cosine distance might mean that they are same person but the previous model needs 0.80 distance value to classify pairs as same person.

**Comment by zul96a** [comment]
Thank you for the quick reply! I noticed that I made an error in my previous comment as it suppose to be true negative, which means the persons are not the same at all.

**Comment by serengil** [comment]
Unfortunately, it is hard to understand.

BTW, I am closing this issue because you are facing with a model out of the deepface.

**Comment by zul96a** [comment]
Sorry if my post caused some confusion.

I have a simple code and some screenshots to give better explanation with my issue.
![image](https://user-images.githubusercontent.com/77947380/135981079-48921ea1-7e05-4c9c-ad58-41242aff61dd.png)

So from the screenshot, image1 and image2 are from person A while image3 is from person B. Using the arcface_weight.h5 provided from this repo, I get this results:
![image](https://user-images.githubusercontent.com/77947380/135979365-13cede89-b6a8-49eb-9fb0-55f49c403524.png)
This is expected as the model can differentiate each of the persons. 

Currently, I'm building a service that takes a selfie of someone and compare it with their ID face photo for verification purpose.
In other words, both of the images must be the same person in order to pass the verification step. I have watched your Youtube
video and after I did some ROC Calculation to determine the performance and suitable threshold (the data I used for the ROC
Calculation comprises of 2000 instances of 400 persons) and I get the threshold to be around 0.51 with AUC score around 0.87.

So at the moment, I'm currently doing Finetuning Arcface model using Keras_Insightface implementation. Using around 4400 persons with around 14k images and by following the steps here https://github.com/leondgarse/Keras_insightface/discussions/36 I manage to fine-tuned a resnet34 model from this repo (the one trained on CASIA) with my set of images. This is the model score after finishing finetuning.
<img width="472" alt="1" src="https://user-images.githubusercontent.com/77947380/135991257-6242fbe9-9213-48e7-b683-62ee113c6e5a.PNG">
After that, I saved the model using model.save_weights just to save its weight. I realize the new model has a slightly bigger size than from this repo 
![image](https://user-images.githubusercontent.com/77947380/135991601-c25e8dd7-6729-4bea-86d1-2769c2fa0587.png)
. I tried to use this model by replace the one inside the /.deepface/weights folder and rename it to arcface_weights.h5. The results were very different and inaccurate at every time I do the verification even using the same codes and images (1st image in this comment). 

I also noticed from the console that the weights of this model aren't loaded properly. 
![image](https://user-images.githubusercontent.com/77947380/135992040-daaf0e62-71c1-4bb4-ab14-460a22c7f7e5.png)

Here are some screenshots for the summary of my model. Here is the top layers
![image](https://user-images.githubusercontent.com/77947380/135992202-bdd7f6cf-3d5e-47f9-9936-a74a7cd92f33.png)
And the bottom layers 
![image](https://user-images.githubusercontent.com/77947380/135992389-befac101-ce48-4d8a-8a6c-df77b201e178.png)

So I think the issue is the model weight can't be loaded properly or might be the issue of tensorflow version?

**Comment by HamDan1999** [comment]
> Hi again!
> 
> Referring to the title, I manage to retrain Arcface model using [Keras Insightface](https://github.com/leondgarse/Keras_insightface) implementation. I looked at arcface model from this repo and based on the model.summary(), it uses Resnet34 as base model.
> 
> Following [leondgarse/Keras_insightface#36](https://github.com/leondgarse/Keras_insightface/discussions/36) method on training Resnet34 model, I use CASIA webface plus another 4200 unique person from my database and I got accuracy of around 0.99% on lfw , 0.89% for agedb_30.
> 
> When I tried to replace the newly trained model, h5 , with the current arface_weights.h5, I see the results vary too much. For example before this, a true negative reading (cosine distance) between Image A and B was 0.89 but after changing the weight, it become 0.11 which is false positive. In your repo model, set the threshold to be around 0.51 which anything below it is match
> 
> I suspect that the newly trained model structure might differ a bit from this repo version since the size in Bytes of the new model is 133,947 KB , instead of this repo, 133, 816. I also notice that from the Keras Insightface repo, they use a custom NormDense layer, so maybe this might be the issue? Or it could be tensorflow version?
> 
> I'm not sure what I can do at the moment, but I appreciate any insights that you can give me. Thank you!

Hello, I hope you both are doing well.

I am really interested in your code and performance you achieved. Could you please share the source code with the results so that I can reproduce them?

Email: mohdalansari99@hotmail.com

Thanks for your time.
